src/core/thread/network_data.cpp

```diff
@@ -284,112 +284,106 @@ Error NetworkData::GetNextServiceId(Iterator &aIterator, uint16_t aRloc16, uint8
+    ExternalRouteConfig route;
 
-    while (aCompare.GetNextExternalRoute(outerIterator, aRloc16, outerConfig) == kErrorNone)
+    while (GetNextExternalRoute(iterator, aRoute.mRloc16, route) == kErrorNone)
     {
-        Iterator            innerIterator = kIteratorInit;
-        ExternalRouteConfig innerConfig;
-        Error               error;
-
-        while ((error = GetNextExternalRoute(innerIterator, aRloc16, innerConfig)) == kErrorNone)
+        if (route == aRoute)
         {
-            if (outerConfig == innerConfig)
-            {
-                break;
-            }
-        }
-
-        if (error != kErrorNone)
-        {
-            ExitNow(rval = false);
+            contains = true;
+            break;
         }
     }
 
-exit:
-    return rval;
+    return contains;
 }
 
-bool NetworkData::ContainsServices(const NetworkData &aCompare, uint16_t aRloc16) const
+bool NetworkData::ContainsService(const ServiceConfig &aService) const
 {
-    Iterator      outerIterator = kIteratorInit;
-    ServiceConfig outerConfig;
-    bool          rval = true;
+    bool          contains = false;
+    Iterator      iterator = kIteratorInit;
+    ServiceConfig service;
 
-    while (aCompare.GetNextService(outerIterator, aRloc16, outerConfig) == kErrorNone)
+    while (GetNextService(iterator, aService.GetServerConfig().mRloc16, service) == kErrorNone)
     {
-        Iterator      innerIterator = kIteratorInit;
-        ServiceConfig innerConfig;
-        Error         error;
-
-        while ((error = GetNextService(innerIterator, aRloc16, innerConfig)) == kErrorNone)
+        if (service == aService)
         {
-            if (outerConfig == innerConfig)
-            {
-                break;
-            }
-        }
-
-        if (error != kErrorNone)
-        {
-            ExitNow(rval = false);
+            contains = true;
+            break;
         }
     }
 
-exit:
-    return rval;
+    return contains;
 }
 
 bool NetworkData::ContainsService(uint8_t aServiceId, uint16_t aRloc16) const
 {
+    bool     contains = false;
     Iterator iterator = kIteratorInit;
     uint8_t  serviceId;
-    bool     rval = false;
 
     while (GetNextServiceId(iterator, aRloc16, serviceId) == kErrorNone)
     {
         if (serviceId == aServiceId)
         {
-            ExitNow(rval = true);
+            contains = true;
+            break;
+        }
+    }
+
+    return contains;
+}
+
+bool NetworkData::ContainsEntriesFrom(const NetworkData &aComapre, uint16_t aRloc16, bool aExcludeService) const
+{
+    bool     contains = true;
+    Iterator iterator = kIteratorInit;
+
+    while (true)
+    {
+        Config              config;
+        OnMeshPrefixConfig  prefix;
+        ExternalRouteConfig route;
+        ServiceConfig       service;
+
+        config.mOnMeshPrefix  = &prefix;
+        config.mExternalRoute = &route;
+        config.mService       = aExcludeService ? nullptr : &service;
```

Comment:
Do we really need to handle this case? If service feature is not enabled on a device, the local network data should never include service TLVs.

Reply:
I was thinking of the same. 

I think the current `main` code, does only check the services between `Local` and `Leader` when `OPENTHREAD_CONFIG_TMF_NETDATA_SERVICE_ENABLE` is enabled, so I tried to keep the behavior the same.

But I agree that it make sense to check the service entries independent of `SERVICE_ENABLE` config. 
- If  `SERVICE_ENABLE` config is not used/enabled, device should not be able to add any service entry to network data, so if we have have any service entries in`Leader` network data under the device's RLOC16, then it is not consistent (device needs to retry registering).

I will update the code and remove the `aExcludeService` (it helps make it even simpler 😃).